anko.go

```diff
@@ -17,75 +16,81 @@ import (
 	"github.com/mattn/anko/vm"
 
 	"github.com/daviddengcn/go-colortext"
-	"github.com/mattn/go-isatty"
 )
 
-const version = "0.0.2"
+const version = "0.0.3"
 
 var (
-	fs   = flag.NewFlagSet(os.Args[0], 1)
-	line = fs.String("e", "", "One line of program")
-	v    = fs.Bool("v", false, "Display version")
-
-	istty = isatty.IsTerminal(os.Stdout.Fd())
+	flagExecute *string
+	env         *vm.Env
 )
 
-func colortext(color ct.Color, bright bool, f func()) {
-	if istty {
-		ct.ChangeColor(color, bright, ct.None, false)
-		f()
-		ct.ResetColor()
-	} else {
-		f()
-	}
+func main() {
+	parseFlags()
+	setupEnv()
+	runNonInteractive()
+	runInteractive()
 }
 
-func main() {
-	fs.Parse(os.Args[1:])
-	if *v {
+func parseFlags() {
+	flagVersion := flag.Bool("v", false, "prints out the version and then exits")
+	flagExecute = flag.String("e", "", "execute the Anko code")
+	flag.Parse()
+
+	if *flagVersion {
 		fmt.Println(version)
 		os.Exit(0)
 	}
+}
 
-	var (
-		code      string
-		b         []byte
-		reader    *bufio.Reader
-		following bool
-		source    string
-	)
-
-	env := vm.NewEnv()
-	interactive := fs.NArg() == 0 && *line == ""
+func setupEnv() {
+	env = vm.NewEnv()
+	if len(*flagExecute) > 0 || flag.NArg() < 1 {
```

Comment:
please use flagExecute != "" instead. len(flagExecute) > 0 will be bit slower.

Reply:
The performance difference is going to be so very small in this case because it is only called once, so I am fine either way. 

If you are really interested, my understanding is len is faster because len is a built in function.

https://groups.google.com/forum/#!topic/golang-nuts/7Ks1iq2s7FA

Reply:
Yes, this is micro benchmark.

![](https://go-gyazo.appspot.com/d7ba8ce5497d6872.png)

```
package main

import (
	"strings"
	"testing"
)

func BenchmarkEqEq(b *testing.B) {
	s := []string{"bar", "", strings.Repeat("o", 9999)}
	n := 0
	for i := 0; i < b.N; i++ {
		if s[i%3] != "" {
			n++
		}
	}
}

func BenchmarkLen(b *testing.B) {
	s := []string{"bar", "", strings.Repeat("o", 9999)}
	n := 0
	for i := 0; i < b.N; i++ {
		if len(s[i%3]) > 0 {
			n++
		}
	}
}
```
This is just small thing for my coding style. :)

Reply:
Going to change it to !=

That being said, found them to be the same speed.

```
func BenchmarkEqEq(b *testing.B) {
	s := make([]string, 10000)
	for i := 0; i < 10000; i++ {
		s[i] = "a"
	}
	n := 0
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if s[n] != "" {
			n++
		}
		if n > 9999 {
			n = 0
		}
	}
}

func BenchmarkLen(b *testing.B) {
	s := make([]string, 10000)
	for i := 0; i < 10000; i++ {
		s[i] = "a"
	}
	n := 0
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if len(s[n]) > 0 {
			n++
		}
		if n > 9999 {
			n = 0
		}
	}
}
```

```
C:\>go test -v -bench=. test
goos: windows
goarch: amd64
pkg: test
BenchmarkEqEq-8         2000000000               1.04 ns/op
BenchmarkLen-8          2000000000               1.04 ns/op
PASS
ok      test    4.425s

C:\>go test -v -bench=. test
goos: windows
goarch: amd64
pkg: test
BenchmarkEqEq-8         2000000000               1.04 ns/op
BenchmarkLen-8          2000000000               1.04 ns/op
PASS
ok      test    4.403s

C:\>go test -v -bench=. test
goos: windows
goarch: amd64
pkg: test
BenchmarkEqEq-8         2000000000               1.04 ns/op
BenchmarkLen-8          2000000000               1.04 ns/op
PASS
ok      test    4.409s
```

Reply:
Oh, also if the string is bigger, same result.

Also can do same test on a single string and take in account caching.

```
func BenchmarkEqEq(b *testing.B) {
	s := "a"
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if s != "" {
		}
	}
}

func BenchmarkLen(b *testing.B) {
	s := "a"
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if len(s) > 0 {
		}
	}
}
```

```
C:\>go test -v -bench=. test
goos: windows
goarch: amd64
pkg: test
BenchmarkEqEq-8         2000000000               0.26 ns/op
BenchmarkLen-8          2000000000               0.26 ns/op
PASS
ok      test    1.138s

C:\>go test -v -bench=. test
goos: windows
goarch: amd64
pkg: test
BenchmarkEqEq-8         2000000000               0.26 ns/op
BenchmarkLen-8          2000000000               0.26 ns/op
PASS
ok      test    1.143s

C:\>go test -v -bench=. test
goos: windows
goarch: amd64
pkg: test
BenchmarkEqEq-8         2000000000               0.26 ns/op
BenchmarkLen-8          2000000000               0.26 ns/op
PASS
ok      test    1.151s
```

Reply:
Compiler might be optimized recently, As far as I remember, Russ or Iant say `!=` is bits faster.